### ISSUE TITLE
make main setup.py work with 'build' subcommand

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1835]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#1835|
|Type|defect|
|Created|2006-06-20 17:31:21Z|

setup.py build apparently doesn't work on the sumo tarball, with errors about importing twisted.python.dist, and this is my guess why: when 'install' is run, it is always run in an order with which 'core' will be installed first, then everything else after it. With build, 'core' doesn't actually get installed, so twisted.python.dist isn't importable for the rest of the modules.

Potential fix: set PYTHONPATH to include TwistedCore during the run of successive subprojects' setup.py commands.

<details><summary>Searchable metadata</summary>

```
trac-id__1835 1835
type__defect defect
reporter__radix radix
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1150824681000000 1150824681000000
changetime__1166925828000000 1166925828000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix set owner to @exarkun|
|-|-|

I think I've fixed it in setup-py-depends-on-core-1835. No tests included. see [#2307](https://github.com/twisted/twisted/issues/2307) and [#2308](https://github.com/twisted/twisted/issues/2308) for the tickets which I have filed in anger after implementing this change.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_2"></a>|@radix commented|
|-|-|

Just to note, these are some of the environments I have manually tested:

1. SVN checkout of Twisted with Combinator,

2. Totally barebones dapper machine without Twisted installed (which is the case this ticket is meant to solve), from an extracted Twisted-2.4.0.tar.bz2 with the setup.py replaced with the one from the branch,

3. Same as #2 but after deleting a random twisted subproject directory from the sumo intsall.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @radix|
|-|-|

This change looks good.

One thing I would like to see is an enumeration of the supported uses of setup.py.  The name itself serves as documentation of the fact that it behaves in every way like a distutils-based installer.  However, since this isn't the case, a list of the uses which are explicitly supported would be useful.  I have Twisted developers in mind as the primary audience for this (as a guide for future changes), but I suppose it would be useful for end users as well.

This could take the form of a comment at the top of setup.py.

On the other hand, maybe this is stupid.  Do it if you think it sounds useful, skip it otherwise.

I tested setup.py in a couple different ways, including an svn checkout of Twisted without Combinator.  It behaved correctly in each case.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

I added a slightly more informative docstring, but according to exarkun you should probably still do the merge.

Please do this ASAP.  I'd like to have more than 2 releases in 2006.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix set status to `closed`|
|-|-|

(In [19055]) Merge setup-py-depends-on-core-1835

Authors: radix, glyph
Reviewer: exarkun
Fixes [#1835](https://github.com/twisted/twisted/issues/1835)

This makes it so "setup.py build" and other commands which don't actually install 
Twisted to an importable location will work with the main setup.py script.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
